packages/store/addon/-private/system/snapshot.ts

```diff
@@ -98,34 +143,33 @@ export default class Snapshot {
    @property record
    @type {Model}
    */
-  get record() {
+  get record(): RecordInstance {
     return this._internalModel.getRecord();
   }
 
-  get _attributes() {
-    let attributes = this.__attributes;
-
-    if (attributes === null) {
-      let record = this.record;
-      attributes = this.__attributes = Object.create(null);
-      let attrs;
+  get _attributes(): Dict<any> {
+    if (this.__attributes !== null) {
+      return this.__attributes;
+    }
+    let record = this.record;
+    const attributes = (this.__attributes = Object.create(null));
+    let attrs: string[];
 
-      if (CUSTOM_MODEL_CLASS) {
-        attrs = Object.keys(this._store._attributesDefinitionFor(this.modelName, this.identifier));
-      } else {
-        attrs = Object.keys(this._store._attributesDefinitionFor(this.modelName, record.id));
-      }
-      if (CUSTOM_MODEL_CLASS) {
-        attrs.forEach(keyName => {
-          if (this.type.isModel) {
-            attributes[keyName] = get(record, keyName);
-          } else {
-            attributes[keyName] = recordDataFor(this._internalModel).getAttr(keyName);
-          }
-        });
-      } else {
-        record.eachAttribute(keyName => (attributes[keyName] = get(record, keyName)));
-      }
+    if (CUSTOM_MODEL_CLASS) {
+      attrs = Object.keys(this._store._attributesDefinitionFor(this.modelName, this.identifier));
+    } else {
+      attrs = Object.keys(this._store._attributesDefinitionFor(this.modelName));
+    }
+    if (CUSTOM_MODEL_CLASS) {
+      attrs.forEach(keyName => {
+        if (schemaIsDSModel(this.type)) {
+          attributes[keyName] = get(record as DSModel, keyName);
+        } else {
+          attributes[keyName] = recordDataFor(this._internalModel).getAttr(keyName);
+        }
+      });
+    } else {
+      (record as DSModel).eachAttribute(keyName => (attributes[keyName] = get(record as DSModel, keyName)));
```

Comment:
Is this cast safe because we're not in the `CUSTOM_MODEL_CLASS`? Can we do it with type narrowing or at a minimum add a comment?

Reply:
It's safe because we are not in that flag branch, there isn't a good way to do it with narrowing because we can't type model yet but I can add a comment.

Reply:
(to clarify, we can't type model && there isn't a good prop for us to be able to narrow by that won't be there for other known alternative models)

Reply:
That's what I figured. Thanks for clarifying!